### ISSUE TITLE
Unable to process Assertion which contains AuthnContextDeclRef 

### ISSUE THREAD

**Issue by hnnweb** [ITfoxtec/ITfoxtec.Identity.Saml2#62]
This error occurs when `AuthnContextDeclRef` is in the response.

Related errors in other projects:
https://github.com/Sustainsys/Saml2/pull/249

Complete stacktrace:
```
Microsoft.IdentityModel.Tokens.Saml2.Saml2SecurityTokenException: IDX13001: A SAML2 assertion that specifies an AuthenticationContext DeclarationReference is not supported.To handle DeclarationReference, extend the Saml2SecurityTokenHandler and override ProcessAuthenticationStatement.
   at Microsoft.IdentityModel.Tokens.Saml2.Saml2SecurityTokenHandler.ProcessAuthenticationStatement(Saml2AuthenticationStatement statement, ClaimsIdentity identity, String issuer)

   at Microsoft.IdentityModel.Tokens.Saml2.Saml2SecurityTokenHandler.ProcessStatements(ICollection`1 statements, ClaimsIdentity identity, String issuer)

   at Microsoft.IdentityModel.Tokens.Saml2.Saml2SecurityTokenHandler.CreateClaimsIdentity(Saml2SecurityToken samlToken, String issuer, TokenValidationParameters validationParameters)

   at ITfoxtec.Identity.Saml2.Tokens.Saml2ResponseSecurityTokenHandler.ValidateToken(SecurityToken token, String tokenString, Saml2Response saml2Response)

   at ITfoxtec.Identity.Saml2.Saml2AuthnResponse.ReadClaimsIdentity(String tokenString)

   at ITfoxtec.Identity.Saml2.Saml2AuthnResponse.Read(String xml, Boolean validateXmlSignature)

   at ITfoxtec.Identity.Saml2.Saml2PostBinding.Read(HttpRequest request, Saml2Request saml2RequestResponse, String messageName, Boolean validateXmlSignature)

   at ITfoxtec.Identity.Saml2.Saml2Binding`1.ReadSamlResponse(HttpRequest request, Saml2Response saml2Response)

   at Example.Controllers.AuthController.AssertionConsumerService() in C:\Users\heno\source\repos\Example\Controllers\AuthController.cs:line 65

   at Microsoft.AspNetCore.Mvc.Infrastructure.ActionMethodExecutor.TaskOfIActionResultExecutor.Execute(IActionResultTypeMapper mapper, ObjectMethodExecutor executor, Object controller, Object[] arguments)

   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeActionMethodAsync>g__Awaited|12_0(ControllerActionInvoker invoker, ValueTask`1 actionResultValueTask)

   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeNextActionFilterAsync>g__Awaited|10_0(ControllerActionInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)

   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Rethrow(ActionExecutedContextSealed context)

   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)

   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.InvokeInnerFilterAsync()

--- End of stack trace from previous location where exception was thrown ---
```

Where row 65 is `binding.ReadSamlResponse(Request.ToGenericHttpRequest(), saml2AuthnResponse);` in:
```
        [Route("AssertionConsumerService")]
        public async Task<IActionResult> AssertionConsumerService()
        {
            var binding = new Saml2PostBinding();
            var saml2AuthnResponse = new Saml2AuthnResponse(config);

            binding.ReadSamlResponse(Request.ToGenericHttpRequest(), saml2AuthnResponse);
            if (saml2AuthnResponse.Status != Saml2StatusCodes.Success)
            {
                throw new AuthenticationException($"SAML Response status: {saml2AuthnResponse.Status}");
            }
            binding.Unbind(Request.ToGenericHttpRequest(), saml2AuthnResponse);
            await saml2AuthnResponse.CreateSession(HttpContext, claimsTransform: (claimsPrincipal) => ClaimsTransform.Transform(claimsPrincipal));

            var relayStateQuery = binding.GetRelayStateQuery();
            var returnUrl = relayStateQuery.ContainsKey(relayStateReturnUrl) ? relayStateQuery[relayStateReturnUrl] : Url.Content("~/");
```

For simplicity the project is this repo and the error:
https://github.com/ITfoxtec/ITfoxtec.Identity.Saml2/blob/master/test/TestWebAppCoreFramework/Controllers/AuthController.cs

**Comment by hnnweb** [comment]
After reading the error once more it seems that 
[Tokens/Saml2ResponseSecurityTokenHandler.cs](https://github.com/ITfoxtec/ITfoxtec.Identity.Saml2/blob/master/src/ITfoxtec.Identity.Saml2/Tokens/Saml2ResponseSecurityTokenHandler.cs) needs to override ProcessAuthenticationStatement

**Comment by Revsgaard** [comment]
Thank you for the error description, I'll look into the problem.

Regards Anders

**Comment by hnnweb** [comment]
The problem seems to appear if the counterpart is running Microfocus:

https://community.microfocus.com/t5/Access-Manager-User-Discussions/Can-you-suppress-AuthnContextDeclRef-in-a-SAML2-assertion/td-p/2229887

The customer in my end is running Microfocus.

**Comment by Revsgaard** [comment]
I have release 4.1.2-beta2 which should resolve the problem https://github.com/ITfoxtec/ITfoxtec.Identity.Saml2/releases/tag/4.1.2-beta2.

Please test and let me know if it solves your problem.

**Comment by hnnweb** [comment]
Looks good

**Comment by Revsgaard** [comment]
Released in version 4.2.0 https://github.com/ITfoxtec/ITfoxtec.Identity.Saml2/releases/tag/4.2.0
